crates/testing/tests/proposal_ordering.rs

```diff
@@ -0,0 +1,110 @@
+use hotshot::{tasks::task_state::CreateTaskState, types::SystemContextHandle};
+use hotshot_example_types::node_types::{MemoryImpl, TestTypes};
+use hotshot_task_impls::{consensus::ConsensusTaskState, events::HotShotEvent::*};
+use hotshot_testing::{
+    predicates::{exact, is_at_view_number, quorum_proposal_send},
+    task_helpers::vid_scheme_from_view_number,
+    view_generator::TestViewGenerator,
+};
+use hotshot_types::{data::ViewNumber, traits::node_implementation::ConsensusTime};
+use jf_primitives::vid::VidScheme;
+
+/// Runs the test specified in this file with a boolean flag that determines whether or not to make
+/// the `QCFormed` event come first in the inputs, or last. Since there's only two possible cases
+/// to check, the code simply swaps their order in the input vector.
+async fn test_ordering_with_specific_order(qc_formed_first: bool) {
+    use hotshot_testing::script::{run_test_script, TestScriptStage};
+    use hotshot_testing::task_helpers::build_system_handle;
+
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+
+    let node_id = 2;
+    let handle = build_system_handle(node_id).await.0;
+    let quorum_membership = handle.hotshot.memberships.quorum_membership.clone();
+
+    let vid =
+        vid_scheme_from_view_number::<TestTypes>(&quorum_membership, ViewNumber::new(node_id));
+
+    // Make some empty encoded transactions, we just care about having a commitment handy for the
+    // later calls.
+    let encoded_transactions = Vec::new();
+    let vid_disperse = vid.disperse(&encoded_transactions).unwrap();
+    let payload_commitment = vid_disperse.commit;
+
+    let mut proposals = Vec::new();
+    let mut votes = Vec::new();
+    let mut leaders = Vec::new();
+
+    let mut generator = TestViewGenerator::generate(quorum_membership.clone());
+
+    for view in (&mut generator).take(3) {
+        proposals.push(view.quorum_proposal.clone());
+        votes.push(view.create_vote(&handle));
+        leaders.push(view.leader_public_key);
+    }
+
+    // This stage transitions from the initial view to view 1
+    let view_0 = TestScriptStage {
+        inputs: vec![QuorumProposalRecv(proposals[0].clone(), leaders[0])],
+        outputs: vec![
+            exact(ViewChange(ViewNumber::new(1))),
+            exact(QuorumProposalValidated(proposals[0].data.clone())),
+            exact(QuorumVoteSend(votes[0].clone())),
+        ],
+        asserts: vec![is_at_view_number(1)],
+    };
+
+    // Node 2 is the leader up next, so we form the QC for it.
+    let cert = proposals[1].data.justify_qc.clone();
+    let view_1_inputs = if qc_formed_first {
+        vec![
+            QuorumProposalRecv(proposals[1].clone(), leaders[1]),
+            QCFormed(either::Left(cert)),
```

Comment:
Not sure how to do this without first receiving a proposal and getting the view change. I _could_ do this in a separate stage, but it would still create the case where QuorumProposalRecv needs to be in a consistent place in the vec (namely, the first).